### PR TITLE
Pass features along during expansion

### LINKED ISSUE #22234: `feature_gate::MacroVisitor` misses macro invocations nested in macro invocations
Today, this program is correctly rejected:
```rust
fn main() {
    unsafe {
        let x = asm!("");
        println!("hi {:?}", x);
    }
}
```
due to a missing `#![feature(asm)]`

Unfortunately, this variant program is accepted by the compiler:
```rust
fn main() {
    unsafe {
        println!("hi {:?}", asm!(""));
    }
}
```

The reason (I think) is that the `fn visit_mac` in `feature_gate::MacroVisitor` needs to recursively invoke the visitor on the expression inputs to the macro.

### PR DESCRIPTION
**Pull request by pnkfelix**
Pass features along during expansion

Use the set of passed features to detect uses of feature-gated macros without the corresponding feature enabled.

Fix #22234.

---

Also, the framework this add (passing along a reference to the features in the expansion context) is a necessary precursor for landing a properly feature-gated desugaring-based overloaded-`box` and placement-`in` (#22181).

---

This is fixing a bug, but since there might be code out there that is unknowingly taking advantage of that bug, I feel obligated to mark this as a:

[breaking-change]

### CODE REVIEW

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -590,11 +609,18 @@ fn check_crate_inner<F>(cm: &CodeMap, span_handler: &SpanHandler, krate: &ast::C
 
     check(&mut cx, krate);
 
+    // FIXME (pnkfelix): Before adding the 99th entry below, change it
+    // to a single-pass (instead of N calls to `.has_feature`)a.
```
Comment (huonw) [inline src/libsyntax/feature_gate.rs]: s/a//

File: src/libsyntax/feature_gate.rs [hunk 2]
```diff
@@ -230,24 +250,23 @@ impl<'a, 'v> Visitor<'v> for MacroVisitor<'a> {
         let ast::MacInvocTT(ref path, _, _) = mac.node;
         let id = path.segments.last().unwrap().identifier;
 
+        // Issue 22234: If you add a new case here, make sure to also
+        // add code to catch the macro during or after expansion.
+
         if id == token::str_to_ident("asm") {
```
Comment (huonw) [inline src/libsyntax/feature_gate.rs]: Theoretically these are redundant now (that is, they'd be caught by expansion), but I guess you have a motivation for keeping them around?

**Comment by pnkfelix** [comment]
ah good old tidy

File: src/libsyntax/feature_gate.rs (hunk 2 above)
Comment (pnkfelix) [inline src/libsyntax/feature_gate.rs]: i think its good to catch the uses before expansion when we can.  i.e. I want to catch the ones hidden by `cfg` switches, when possible.

Comment (huonw) [inline src/libsyntax/feature_gate.rs]: Could you add a short comment mentioning this?

**Comment by huonw** [comment]
r=me with the comment and maybe some squashing.

**Pull request merged**